modify travis tests to create user with password for mysql and postgres

dennisobrien opened the pull request:
Modify the Travis CI tests to create a database user with a password and use these in the SQLALCHEMY_DATABASE_URI for the tests for mysql and postgres.

This is related to #1070 and I expect this PR to have failing Travis tests.

It's fine to reject this PR -- I just wanted to verify that the tests will fail.

dennisobrien committed:
modify travis tests to create user with password for mysql and postgres and use these in the SQLALCHEMY_DATABASE_URI

xrmx commented:
@dennisobrien please do experiments on your own fork instead of here

mistercrunch committed:
[sqllab] hide SqlEditor textarea to prevent flicker

vera-liu committed:
Single quote filter values with comma (#1084)

* Single quote filter values with comma

* refactor for codeclimate limite

* Added unit tests and tooltip

mistercrunch committed:
[sql lab] specify schema name when generating vanila query (#1096)

* [sql lab] specify schema name when generating vanila query

* Fixing some react warnings

yejianye committed:
Time grain support for unix-timestamp columns (#1093)

* Add time grain support for time columnd in unix timestamp

* Fix datetime parsing for unix epoch

Since we've already converted unix epoch to datetime type,
we shouldn't specify 'unit' parameter in pandas.to_datetime

* Fix SQLite timestamp to datetime conversion

bkyryliuk committed:
Clean up the druid sync api. (#1101)

mistercrunch committed:
[bugfix] filter_immune_slices doesn't work (#1110)

dennisobrien committed:
added set_sqlalchemy_uri which stores a password-masked version in the 'sqlalchemy_uri' field and the encrypted password in the 'password' field.
Modified 'testconn' to work with this.  If the passed uri is different from the masked uri, use the passed uri.  Otherwise, use the decrypted uri associated with this database.

dennisobrien committed:
added set_sqlalchemy_uri which stores a password-masked version in the 'sqlalchemy_uri' field and the encrypted password in the 'password' field.
Modified 'testconn' to work with this.  If the passed uri is different from the masked uri, use the passed uri.  Otherwise, use the decrypted uri associated with this database.

dennisobrien committed:
added set_sqlalchemy_uri which stores a password-masked version in the 'sqlalchemy_uri' field and the encrypted password in the 'password' field.
Modified 'testconn' to work with this.  If the passed uri is different from the masked uri, use the passed uri.  Otherwise, use the decrypted uri associated with this database.

dennisobrien committed:
broke the comment into two lines to make linting happy

dennisobrien committed:
fixed trailing whitespace

mistercrunch committed:
Animated GIFs on README (#1109)

dennisobrien committed:
fixed a spelling mistake and removed a commented line.

dennisobrien commented:
The original intent was to test SQLALCHEMY_DATABASE_URI with a username and password as a test for #1070 (sorry @xrmx -- I only realized after that I could have run the same tests Travis runs on my own machine).

The follow up commits address #1070 
- At initialization `utils.get_or_create_main_db` calls `dbobj.set_sqlalchemy_uri(config.get("SQLALCHEMY_DATABASE_URI"))`
- `caravel.models.Database.set_sqlalchemy_uri` stores the password-masked uri in the field `sqlalchemy_uri` and the encrypted password in `password`.
- That logic was removed from `caravel.views.DatabaseView.pre_add`.
    - There is still one line in `pre_add` that I did not move because I don't know where it belongs: `utils.merge_perm(sm, 'database_access', db.perm)`
- This refactor broke `testconn` so I made some changes here:
    - Include `database_name` in the json payload.
    - If `uri` is the same as `database.safe_sqlalchemy_uri()`, use the stored `database.sqlalchemy_uri_decrypted()`.
    - Otherwise, assume the user is trying a different connection string and test that.
    - I really don't know if this is an important feature but it's probably good to respect the user's intent and give accurate feedback.

mistercrunch committed:
[bugfix] namespacing the mapbox css

xrmx commented:
@dennisobrien  If you are going to poke with testconn please cherry-pick whatever seems sensible (the test at least) from this pull request:
https://github.com/airbnb/caravel/pull/694

dennisobrien committed:
Incorporate the tests from PR #694 which addresses #596

dennisobrien commented:
Thanks @xrmx -- I wasn't aware of that test.  I brought in the test but I didn't use git cherry-pick since I needed just part of the commit.  I still don't have tests running properly on my dev machine so I'm committing the changes with the hopes that the tests will pass.

Do you recommend that this PR supercedes #694 or that I merge that branch into this one and deal with the merge conflicts?

xrmx commented:
@dennisobrien no need to do that, i am going to close #694 if all the relevant pieces are included there :)

xrmx commented on tests/core_tests.py:
```diff
@@ -114,6 +114,34 @@ def test_misc(self):
+        response = self.client.post('/caravel/testconn', data=data, content_type='application/json')
+        assert response.status_code == 200
+
+        database = (
```
have you checked this return something with a password?

dennisobrien replied:
This test has three parts, but now that I think of it, it should really just use the last two since some of the test cases are against a sqlite db and others use mysql or postgres.  So I think I should remove the hard-coded uri here and just rely on the other two tests.  One of those two tests with the password masked, the other with the password included in the uri.

dennisobrien committed:
handle the case when a uri is passed but a database name is not

dennisobrien committed:
remove the test using the hard-coded sqlalchemy uri.
This leaves two tests: one that uses the password-masked uri, and the other that uses the uri with the password included.

xrmx commented on caravel/views.py:
```diff
@@ -1241,21 +1241,25 @@ def add_slices(self, dashboard_id):
-                .filter_by(database_name=database_name)
-                .first()
-            )
             request_uri = request.json.get('uri')
```
uri = request.json.get('uri')

so that you can save a couple of assigments later?

dennisobrien replied:
I did some refactoring here just as you posted this comment.  I could skip creating `request_uri` but I thought I'd be explicit.

dennisobrien replied:
I simplified the forking paths here in 8c91119
Thanks for the advice!

dennisobrien commented:
Hmm, I'm getting a new test failure that has me scratching my head.  This passed in the previous Travis run so I'm not sure what's happening here.

```
======================================================================
ERROR: test_run_async_query (tests.celery_tests.CeleryTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/airbnb/caravel/tests/celery_tests.py", line 235, in test_run_async_query
    df1 = pd.read_sql_query(query1.select_sql, con=eng)
  File "/home/travis/build/airbnb/caravel/.tox/py27-postgres/lib/python2.7/site-packages/pandas/io/sql.py", line 431, in read_sql_query
    parse_dates=parse_dates, chunksize=chunksize)
  File "/home/travis/build/airbnb/caravel/.tox/py27-postgres/lib/python2.7/site-packages/pandas/io/sql.py", line 1190, in read_query
    result = self.execute(*args)
  File "/home/travis/build/airbnb/caravel/.tox/py27-postgres/lib/python2.7/site-packages/pandas/io/sql.py", line 1081, in execute
    return self.connectable.execute(*args, **kwargs)
  File "/home/travis/build/airbnb/caravel/.tox/py27-postgres/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1991, in execute
    return connection.execute(statement, *multiparams, **params)
  File "/home/travis/build/airbnb/caravel/.tox/py27-postgres/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 912, in execute
    type(object))
InvalidRequestError: Unexecutable object type: <type 'NoneType'>
-------------------- >> begin captured logging << --------------------
flask_appbuilder.security.sqla.manager: INFO: Updated user admin  user
--------------------- >> end captured logging << ---------------------
```

dennisobrien committed:
simplify the logic in testconn.  thanks @xrmx

vera-liu committed:
When the label size is too short, the constant for calculating (#1120)

margin_size does not apply. Also nvd3 auto-adjust font-size of axis
labels.
Temporary solution here: Setting a fixed font-size on nvd3 axis labels
and a minimum threshold for label size.

vera-liu committed:
Associate version to entry files (#1060)

* Associate version to entry files

* Modified path joins for configs

* Made changes based on comments

mistercrunch committed:
[explore] clarifying force-refresh message

mistercrunch committed:
Fixing druid culster perms to mirror sqla databases (#1123)

mistercrunch committed:
[welcome] removing calendar heatmap from welcome page (#1119)

mistercrunch committed:
[hotfix] lint

vera-liu committed:
Apply stretchMargin only to distribution bar (#1124)

* When the label size is too short, the constant for calculating
margin_size does not apply. Also nvd3 auto-adjust font-size of axis
labels.
Temporary solution here: Setting a fixed font-size on nvd3 axis labels
and a minimum threshold for label size.

* Only stretch margin for dist_bar

mistercrunch committed:
[explore] giving more room to Slice title (#1118)

* [explore] giving more room to Slice title

* h2->h3 for slice title

mistercrunch committed:
[security] prevent XSS on FAB list views (#1125)

* [security] prevent XSS on FAB list views

* addressing comments

dennisobrien commented:
The remaining questions I have:
- Why is `DatabaseView.pre_add` not called?  Is it called for others?  Was it called at one point then some dependency changed?
- `DatabaseView.pre_add` is still there but only has a call to `utils.merge_perm`.  Should this be moved to `utils.get_or_create_main_db`?

xrmx commented:
Il 17/09/2016 02:16, dennisobrien ha scritto:
> The remaining questions I have:
>
>   * Why is |DatabaseView.pre_add| not called? Is it called for others? Was it
>     called at one point then some dependency changed?

I've upgraded my test caravel instance to a recent master and it broke 
connections to the database, updating my connection uri and saving the database 
model fixed the issue. So it looks like the hooks works fine before the password 
is clear but not if it's masked?

xrmx committed:
Fix double escaping of dttm expressions (#744) (#1103)

If ddtm_expr is an expression with special characters then timestamp_grain escapes
the special characters already escaped.

Solution discussed with sqlalchemy upstream:
https://bitbucket.org/zzzeek/sqlalchemy/issues/3737/literal_column-given-a-specific-sql

Fix #617

xrmx committed:
Make sql test connection work with saved Database instance (#694)

Fix #596

dennisobrien commented:
#694 was merged into master.  I'm dealing with the merge conflicts in this branch and will push something shortly.

dennisobrien committed:
added a newline

dennisobrien committed:
fixing this comment line again

dennisobrien commented:
I've found myself in an endless loop of git rebase in order to try to deal with these branch conflicts but I'm afraid I'm making a bigger mess.  If it is simpler to close this PR and have me open another I'm happy to do that.

xrmx commented:
@dennisobrien do whatever you feel more comfortable with :)

dennisobrien commented:
Closing this PR and starting a clean PR.
